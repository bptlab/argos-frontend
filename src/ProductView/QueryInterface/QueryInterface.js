import React, { Component } from 'react';
import Modal from './Modal/Modal.js';
import QueryInterfaceHeader from './QueryInterfaceHeader/QueryInterfaceHeader.js';

class QueryInterface extends Component {
    constructor(props) {
        super(props);
        this.state = {
            eventTypeName: '',
            eventTypeAttributes: [
                {
                    "id": 0,
                    "name": "timestamp",
                    "type": "Date",
                    "readonly": true
                },
                {
                    "id": 1,
                    "name": "productId",
                    "type": "Integer",
                    "readonly": true
                },
                {
                    "id": 2,
                    "name": "productFamilyId",
                    "type": "String",
                    "readonly": true
                }
            ],
            eventQuery: '',
            validationResult: ''
        };
        this.handleChangeAttributeName = this.handleChangeAttributeName.bind(this);
        this.handleChangeAttributeType = this.handleChangeAttributeType.bind(this);
        this.handleChangeQuery = this.handleChangeQuery.bind(this);
        this.handleSaveQuery = this.handleSaveQuery.bind(this);
    }

    componentDidMount() {
        this.setState({ query: this.props.defaultQuery });
    }

    handleChangeAttributeName(id, name) {
        const attributes = this.state.eventTypeAttributes;
        for (let i = 0; i < this.state.eventTypeAttributes.length; i++) {
            if (attributes[i].id === id && attributes[i].readonly === false) {
                attributes[i].name = name;
                this.setState({ eventTypeAttributes: attributes });
            }
        }
    }

    handleChangeAttributeType(id, type) {
        const attributes = this.state.eventTypeAttributes;
        for (let i = 0; i < this.state.eventTypeAttributes.length; i++) {
            if (attributes[i].id === id && attributes[i].readonly === false) {
                attributes[i].type = type;
                this.setState({ eventTypeAttributes: attributes });
            }
        }
    }

    handleChangeQuery(event) {
        this.setState({ query: event.target.value });
    }

    handleSaveQuery() {
        try {
            this.validateEventType();
            this.validateQuery();
        }
        catch (e) {
            return this.setState({ validationResult: e });
        }
        const attributes = {};
        for (let i = 0; i < this.state.eventTypeAttributes.length; i++) {
            if (!this.state.eventTypeAttributes[i].name === ''){
                attributes[this.state.eventTypeAttributes[i].name] = this.state.eventTypeAttributes[i].type;
            }
        }
        const eventType = {
            'name': this.state.eventTypeName,
            'timestamp': 'timestamp',
            'attributes': attributes
        };
        this.props.dataSource.createEventtype(this.state.eventQuery, eventType, this.handleEventTypeData, this.handleError);
    }

    validateEventType() {
    }

    validateQuery() {
        const insertInto = /\s+INSERT\s+INTO\s+(\w*)/i;
        const match = insertInto.exec(this.state.eventQuery);
        if (match[1] !== this.state.eventTypeName) {
            throw '"' + match[1] + '" has to match the EventType name.';
        }
    handleChangeQuery(event) {
        this.setState({ eventQuery: event.target.value });
    }

    render() {
        return (
            <div className="query-interface">
                <Modal title="Create new query" onSave={this.handleSaveQuery}>
                    <QueryInterfaceHeader eventTypeName={this.state.eventTypeName}
                                          onChangeEventTypeName={this.handleChangeEventTypeName}
                                          eventTypeAttributes={this.state.eventTypeAttributes}
                                          onChangeAttributeName={this.handleChangeAttributeName}
                                          onChangeAttributeType={this.handleChangeAttributeType} />
                    <div className="form-group">
                        <label htmlFor="event-query" className="form-control-label">Event Query</label>
                        <textarea type="text" className="form-control" id="event-query" rows="8"
                                  value={this.state.query}
                                  onChange={this.handleChangeQuery}/>
                    </div>
                    <p>{this.state.validationResult}</p>
                </Modal>
            </div>
        );
    }
} export default QueryInterface;