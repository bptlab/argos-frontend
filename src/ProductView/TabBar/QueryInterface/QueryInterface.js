import React, { Component } from 'react';
import Modal from './Modal/Modal.js';
import QueryInterfaceHeader from './QueryInterfaceHeader/QueryInterfaceHeader.js';

class QueryInterface extends Component {
    constructor(props) {
        super(props);
        this.state = {
            eventTypeName: '',
            eventTypeAttributes: [
                {
                    "id": 0,
                    "name": "productId",
                    "type": "INTEGER",
                    "readonly": true
                },
                {
                    "id": 1,
                    "name": "productFamilyId",
                    "type": "STRING",
                    "readonly": true
                },
                {
                    "id": 2,
                    "name": "",
                    "type": "",
                    "readonly": false
                }
            ],
            eventQuery: '',
            validationResult: '',
            errorMessage: '',
        };
        this.nextAttributeId = 3;
        this.handleChangeEventTypeName = this.handleChangeEventTypeName.bind(this);
        this.handleChangeAttributeName = this.handleChangeAttributeName.bind(this);
        this.handleChangeAttributeType = this.handleChangeAttributeType.bind(this);
        this.handleChangeQuery = this.handleChangeQuery.bind(this);
        this.handleSaveQuery = this.handleSaveQuery.bind(this);
        this.handleSaveQuerySuccess = this.handleSaveQuerySuccess.bind(this);
        this.handleSaveQueryError = this.handleSaveQueryError.bind(this);
    }

    static getDefaultQuery() {
        return "INSERT INTO TestErrorEvents SELECT timestamp, productId, productFamilyId FROM FeedbackData";
    }

    componentDidMount() {
        this.setState({ eventQuery: QueryInterface.getDefaultQuery() });
    }

    addEmptyAttribute(attributes) {
        attributes = attributes.concat([{
            'id': this.nextAttributeId,
            "name": "",
            "type": "",
            "readonly": false
        }]);
        this.nextAttributeId = this.nextAttributeId + 1;
        return attributes;
    }

    getEventType() {
        const eventTypeAttributes = {};
        const attributes = this.state.eventTypeAttributes;
        for (let i = 0; i < attributes.length; i++) {
            if (attributes[i].name) {
                eventTypeAttributes[attributes[i].name] = attributes[i].type;
            }
        }
        const eventType = {
            'name': this.state.eventTypeName,
            'timestamp': 'timestamp',
            'attributes': eventTypeAttributes
        };
        return eventType;
    }

    getEventTypeAttribute(id) {
        return this.state.eventTypeAttributes.find((attribute) => {
            return attribute.id === id;
        });
    }

    handleChangeEventTypeName(name) {
        this.setState({ eventTypeName: name });
    }

    handleChangeAttributeName(id, name) {
        const currentAttribute = this.getEventTypeAttribute(id);
        if (currentAttribute.readonly) {
            return;
        }

        let attributes = this.state.eventTypeAttributes;
        const currentIndex = attributes.indexOf(currentAttribute);
        attributes[currentIndex].name = name;

        if (!name) {
            attributes.splice(currentIndex, 1);
        }
        else if (currentIndex === attributes.length - 1) {
            attributes = this.addEmptyAttribute(attributes);
        }
        this.setState({ eventTypeAttributes: attributes });
    }

    handleChangeAttributeType(id, type) {
        const currentAttribute = this.getEventTypeAttribute(id);
        if (currentAttribute.readonly) {
            return;
        }

        const attributes = this.state.eventTypeAttributes;
        const currentIndex = attributes.indexOf(currentAttribute);
        attributes[currentIndex].type = type;
        this.setState({ eventTypeAttributes: attributes });
    }

    handleChangeQuery(event) {
        this.setState({ eventQuery: event.target.value });
    }

    handleSaveQuery() {
        const eventType = this.getEventType();
        this.props.dataSender.createEventtype(this.state.eventQuery, eventType, this.handleSaveQuerySuccess, this.handleSaveQueryError);
    }

    handleSaveQuerySuccess() {
        $('#query-interface-modal').modal('hide');
    }

    handleSaveQueryError(error) {
        this.setState({ errorMessage: error });
    }

    render() {
        return (
            <div className="query-interface">
                <Modal title="Query Interface" onSave={this.handleSaveQuery}>
                    <QueryInterfaceHeader eventTypeName={this.state.eventTypeName}
                                          onChangeEventTypeName={this.handleChangeEventTypeName}
                                          eventTypeAttributes={this.state.eventTypeAttributes}
                                          onChangeAttributeName={this.handleChangeAttributeName}
                                          onChangeAttributeType={this.handleChangeAttributeType} />
                    <div className="form-group">
                        <label htmlFor="event-query" className="form-control-label">Event Query</label>
                        <textarea type="text" className="form-control"
                                  id="event-query" rows="8"
                                  value={this.state.eventQuery}
                                  onChange={this.handleChangeQuery}/>
                    </div>
                    <p className="error-message">{this.state.errorMessage}</p>
                </Modal>
            </div>
        );
    }
} export default QueryInterface;