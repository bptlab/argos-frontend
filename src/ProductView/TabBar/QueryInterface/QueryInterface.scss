@import "QueryInterfaceAttributeList/QueryInterfaceAttributeList";

.nav-link {
  border-top-radius: 2px !important;
}

#new-complex-eventtype {
  cursor: pointer;
}

.tooltip {
  font-size: 12px;
  @include media-breakpoint-up(uhd){
    font-size: 40px;
  }
}

#event-query{
  border-radius: 0px;
}