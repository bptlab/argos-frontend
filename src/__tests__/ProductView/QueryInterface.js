import React from 'react';
import QueryInterface from '../../ProductView/TabBar/QueryInterface/QueryInterface.js';
import renderer from 'react-test-renderer';

let instance, component;
const callbackMock = jest.fn();

beforeEach(() => {
    const dataSender = {'createEventtype': callbackMock};
    component = renderer.create(
        <QueryInterface
            ref={(child) => {
                instance = child
            }}
            dataSender={dataSender}/>
    );
});

test('Handling change query', () => {
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
    const event = {target: {value: 'TEST'}};
    instance.handleChangeQuery(event);
    tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test('Handling change attribute type', () => {
    instance.handleChangeEventTypeName('Test');
    expect(instance.state.eventTypeName).toBe('Test');

});

test('Handling change attribute type', () => {
    instance.handleChangeAttributeName(0, 'Test');
    expect(instance.state.eventTypeAttributes[0].name).not.toBe('Test');
    instance.handleChangeAttributeName(2, 'Test');
    expect(instance.getEventTypeAttribute(2).name).toBe('Test');
    expect(instance.state.eventTypeAttributes.length).toBe(4);
    instance.handleChangeAttributeName(2, '');
    expect(instance.state.eventTypeAttributes.length).toBe(3);
});

test('Handling change attribute type', () => {
    instance.handleChangeAttributeType(0, 'STRING');
    expect(instance.state.eventTypeAttributes[0].type).not.toBe('STRING');
    instance.handleChangeAttributeType(2, 'INTEGER');
    expect(instance.state.eventTypeAttributes[2].type).toBe('INTEGER');

});

test('Handling save query', () => {
    instance.handleSaveQuery();
    expect(callbackMock).toBeCalled();
});