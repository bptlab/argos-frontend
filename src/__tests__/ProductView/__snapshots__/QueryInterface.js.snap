exports[`test Handling change query 1`] = `
<div
  className="query-interface">
  <a
    className="nav-link"
    data-placement="top"
    data-target="#query-interface-modal"
    data-toggle="modal"
    id="new-complex-eventtype"
    title="Click to create new event type">
    <i
      className="fa fa-plus" />
  </a>
  <div
    className="query-interface modal fade"
    id="query-interface-modal"
    role="dialog"
    tabIndex="-1">
    <div
      className="modal-dialog modal-lg"
      role="document">
      <div
        className="modal-content">
        <div
          className="modal-header">
          <h5
            className="modal-title"
            id="query-interface-label">
            Query Interface
          </h5>
          <button
            className="close"
            data-dismiss="modal"
            type="button">
            <span
              id="close">
              ×
            </span>
          </button>
        </div>
        <div
          className="modal-body">
          
          <div
            className="query-interface-header">
            <div
              className="form-group has-danger">
              <input
                className="form-control"
                id="query-name"
                onChange={[Function]}
                placeholder="Name"
                type="text"
                value="" />
              <div
                className="form-control-feedback">
                This field can\'t be empty.
              </div>
            </div>
            <div
              className="query-interface-header-attribute form-group row">
              <div
                className="input-field input-group col-6">
                <input
                  className="form-control"
                  id="example-text-input"
                  onChange={[Function]}
                  placeholder="Attribute name"
                  readOnly={true}
                  type="text"
                  value="productId" />
              </div>
              <div
                className="input-field input-group col-6">
                <select
                  className="form-control qi-select"
                  onChange={[Function]}
                  readOnly={true}
                  value="INTEGER">
                  <option>
                    STRING
                  </option>
                  <option>
                    BOOLEAN
                  </option>
                  <option>
                    INTEGER
                  </option>
                  <option>
                    FLOAT
                  </option>
                </select>
              </div>
            </div>
            <div
              className="query-interface-header-attribute form-group row">
              <div
                className="input-field input-group col-6">
                <input
                  className="form-control"
                  id="example-text-input"
                  onChange={[Function]}
                  placeholder="Attribute name"
                  readOnly={true}
                  type="text"
                  value="productFamilyId" />
              </div>
              <div
                className="input-field input-group col-6">
                <select
                  className="form-control qi-select"
                  onChange={[Function]}
                  readOnly={true}
                  value="STRING">
                  <option>
                    STRING
                  </option>
                  <option>
                    BOOLEAN
                  </option>
                  <option>
                    INTEGER
                  </option>
                  <option>
                    FLOAT
                  </option>
                </select>
              </div>
            </div>
            <div
              className="query-interface-header-attribute form-group row">
              <div
                className="input-field input-group col-6">
                <input
                  className="form-control"
                  id="example-text-input"
                  onChange={[Function]}
                  placeholder="Attribute name"
                  readOnly={false}
                  type="text"
                  value="" />
              </div>
              <div
                className="input-field input-group col-6">
                <select
                  className="form-control qi-select"
                  onChange={[Function]}
                  readOnly={false}
                  value="">
                  <option>
                    STRING
                  </option>
                  <option>
                    BOOLEAN
                  </option>
                  <option>
                    INTEGER
                  </option>
                  <option>
                    FLOAT
                  </option>
                </select>
              </div>
            </div>
          </div>
          <div
            className="form-group ">
            <label
              className="form-control-label"
              htmlFor="event-query">
              Event Query
            </label>
            <textarea
              className="form-control"
              id="event-query"
              onChange={[Function]}
              rows="8"
              type="text"
              value="INSERT INTO TestErrorEvents SELECT timestamp, productId, productFamilyId FROM FeedbackData" />
            
          </div>
        </div>
        <div
          className="modal-footer">
          <button
            className="btn"
            disabled="disabled"
            onClick={[Function]}
            type="button">
            <div>
              Save
            </div>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`test Handling change query 2`] = `
<div
  className="query-interface">
  <a
    className="nav-link"
    data-placement="top"
    data-target="#query-interface-modal"
    data-toggle="modal"
    id="new-complex-eventtype"
    title="Click to create new event type">
    <i
      className="fa fa-plus" />
  </a>
  <div
    className="query-interface modal fade"
    id="query-interface-modal"
    role="dialog"
    tabIndex="-1">
    <div
      className="modal-dialog modal-lg"
      role="document">
      <div
        className="modal-content">
        <div
          className="modal-header">
          <h5
            className="modal-title"
            id="query-interface-label">
            Query Interface
          </h5>
          <button
            className="close"
            data-dismiss="modal"
            type="button">
            <span
              id="close">
              ×
            </span>
          </button>
        </div>
        <div
          className="modal-body">
          
          <div
            className="query-interface-header">
            <div
              className="form-group has-danger">
              <input
                className="form-control"
                id="query-name"
                onChange={[Function]}
                placeholder="Name"
                type="text"
                value="" />
              <div
                className="form-control-feedback">
                This field can\'t be empty.
              </div>
            </div>
            <div
              className="query-interface-header-attribute form-group row">
              <div
                className="input-field input-group col-6">
                <input
                  className="form-control"
                  id="example-text-input"
                  onChange={[Function]}
                  placeholder="Attribute name"
                  readOnly={true}
                  type="text"
                  value="productId" />
              </div>
              <div
                className="input-field input-group col-6">
                <select
                  className="form-control qi-select"
                  onChange={[Function]}
                  readOnly={true}
                  value="INTEGER">
                  <option>
                    STRING
                  </option>
                  <option>
                    BOOLEAN
                  </option>
                  <option>
                    INTEGER
                  </option>
                  <option>
                    FLOAT
                  </option>
                </select>
              </div>
            </div>
            <div
              className="query-interface-header-attribute form-group row">
              <div
                className="input-field input-group col-6">
                <input
                  className="form-control"
                  id="example-text-input"
                  onChange={[Function]}
                  placeholder="Attribute name"
                  readOnly={true}
                  type="text"
                  value="productFamilyId" />
              </div>
              <div
                className="input-field input-group col-6">
                <select
                  className="form-control qi-select"
                  onChange={[Function]}
                  readOnly={true}
                  value="STRING">
                  <option>
                    STRING
                  </option>
                  <option>
                    BOOLEAN
                  </option>
                  <option>
                    INTEGER
                  </option>
                  <option>
                    FLOAT
                  </option>
                </select>
              </div>
            </div>
            <div
              className="query-interface-header-attribute form-group row">
              <div
                className="input-field input-group col-6">
                <input
                  className="form-control"
                  id="example-text-input"
                  onChange={[Function]}
                  placeholder="Attribute name"
                  readOnly={false}
                  type="text"
                  value="" />
              </div>
              <div
                className="input-field input-group col-6">
                <select
                  className="form-control qi-select"
                  onChange={[Function]}
                  readOnly={false}
                  value="">
                  <option>
                    STRING
                  </option>
                  <option>
                    BOOLEAN
                  </option>
                  <option>
                    INTEGER
                  </option>
                  <option>
                    FLOAT
                  </option>
                </select>
              </div>
            </div>
          </div>
          <div
            className="form-group ">
            <label
              className="form-control-label"
              htmlFor="event-query">
              Event Query
            </label>
            <textarea
              className="form-control"
              id="event-query"
              onChange={[Function]}
              rows="8"
              type="text"
              value="TEST" />
            
          </div>
        </div>
        <div
          className="modal-footer">
          <button
            className="btn"
            disabled="disabled"
            onClick={[Function]}
            type="button">
            <div>
              Save
            </div>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`test Handling save query failure 1`] = `
<div
  className="query-interface">
  <a
    className="nav-link"
    data-placement="top"
    data-target="#query-interface-modal"
    data-toggle="modal"
    id="new-complex-eventtype"
    title="Click to create new event type">
    <i
      className="fa fa-plus" />
  </a>
  <div
    className="query-interface modal fade"
    id="query-interface-modal"
    role="dialog"
    tabIndex="-1">
    <div
      className="modal-dialog modal-lg"
      role="document">
      <div
        className="modal-content">
        <div
          className="modal-header">
          <h5
            className="modal-title"
            id="query-interface-label">
            Query Interface
          </h5>
          <button
            className="close"
            data-dismiss="modal"
            type="button">
            <span
              id="close">
              ×
            </span>
          </button>
        </div>
        <div
          className="modal-body">
          
          <div
            className="query-interface-header">
            <div
              className="form-group has-danger">
              <input
                className="form-control"
                id="query-name"
                onChange={[Function]}
                placeholder="Name"
                type="text"
                value="" />
              <div
                className="form-control-feedback">
                This field can\'t be empty.
              </div>
            </div>
            <div
              className="query-interface-header-attribute form-group row">
              <div
                className="input-field input-group col-6">
                <input
                  className="form-control"
                  id="example-text-input"
                  onChange={[Function]}
                  placeholder="Attribute name"
                  readOnly={true}
                  type="text"
                  value="productId" />
              </div>
              <div
                className="input-field input-group col-6">
                <select
                  className="form-control qi-select"
                  onChange={[Function]}
                  readOnly={true}
                  value="INTEGER">
                  <option>
                    STRING
                  </option>
                  <option>
                    BOOLEAN
                  </option>
                  <option>
                    INTEGER
                  </option>
                  <option>
                    FLOAT
                  </option>
                </select>
              </div>
            </div>
            <div
              className="query-interface-header-attribute form-group row">
              <div
                className="input-field input-group col-6">
                <input
                  className="form-control"
                  id="example-text-input"
                  onChange={[Function]}
                  placeholder="Attribute name"
                  readOnly={true}
                  type="text"
                  value="productFamilyId" />
              </div>
              <div
                className="input-field input-group col-6">
                <select
                  className="form-control qi-select"
                  onChange={[Function]}
                  readOnly={true}
                  value="STRING">
                  <option>
                    STRING
                  </option>
                  <option>
                    BOOLEAN
                  </option>
                  <option>
                    INTEGER
                  </option>
                  <option>
                    FLOAT
                  </option>
                </select>
              </div>
            </div>
            <div
              className="query-interface-header-attribute form-group row">
              <div
                className="input-field input-group col-6">
                <input
                  className="form-control"
                  id="example-text-input"
                  onChange={[Function]}
                  placeholder="Attribute name"
                  readOnly={false}
                  type="text"
                  value="" />
              </div>
              <div
                className="input-field input-group col-6">
                <select
                  className="form-control qi-select"
                  onChange={[Function]}
                  readOnly={false}
                  value="">
                  <option>
                    STRING
                  </option>
                  <option>
                    BOOLEAN
                  </option>
                  <option>
                    INTEGER
                  </option>
                  <option>
                    FLOAT
                  </option>
                </select>
              </div>
            </div>
          </div>
          <div
            className="form-group ">
            <label
              className="form-control-label"
              htmlFor="event-query">
              Event Query
            </label>
            <textarea
              className="form-control"
              id="event-query"
              onChange={[Function]}
              rows="8"
              type="text"
              value="INSERT INTO TestErrorEvents SELECT timestamp, productId, productFamilyId FROM FeedbackData" />
            
          </div>
        </div>
        <div
          className="modal-footer">
          <button
            className="btn"
            disabled="disabled"
            onClick={[Function]}
            type="button">
            <div>
              Save
            </div>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
